cpp/src/arrow/compute/exec/sink_node.cc

```diff
@@ -232,6 +232,77 @@ class ConsumingSinkNode : public ExecNode {
   std::shared_ptr<SinkNodeConsumer> consumer_;
 };
 
+/**
+ * @brief This node is an extension on ConsumingSinkNode
+ * to facilitate to get the output from an execution plan
+ * as a table. We define a custom SinkNodeConsumer to
+ * enable this functionality.
+ */
+
+struct TableSinkNodeConsumer : public arrow::compute::SinkNodeConsumer {
+ public:
+  TableSinkNodeConsumer(std::shared_ptr<Table>* out,
+                        std::shared_ptr<Schema> output_schema, MemoryPool* pool,
+                        Future<> finish)
+      : out_(out),
+        output_schema_(output_schema),
+        pool_(pool),
+        finish_(std::move(finish)) {}
+
+  Status Consume(ExecBatch batch) override {
+    ARROW_ASSIGN_OR_RAISE(auto rb, batch.ToRecordBatch(output_schema_, pool_));
+    if (rb) {
+      batch_vector.push_back(rb);
+    } else {
+      return Status::Invalid("Invalid ExecBatch consumed");
+    }
+    return Status::OK();
+  }
+
+  Future<> Finish() override {
+    ARROW_ASSIGN_OR_RAISE(auto table, Table::FromRecordBatches(batch_vector));
+    *out_ = table;
+    return finish_;
+  }
+
+ private:
+  std::shared_ptr<Table>* out_;
+  std::shared_ptr<Schema> output_schema_;
+  MemoryPool* pool_;
+  Future<> finish_;
+  std::vector<std::shared_ptr<RecordBatch>> batch_vector;
+};
+
+static std::shared_ptr<SinkNodeConsumer> MakeTableSinkConsumer(
+    std::shared_ptr<Table>* out, std::shared_ptr<Schema> output_schema, MemoryPool* pool,
+    Future<> finish) {
+  auto tb_consumer =
+      std::make_shared<TableSinkNodeConsumer>(out, output_schema, pool, finish);
+  return std::move(tb_consumer);
+}
+
+class TableConsumingSinkNode : public ConsumingSinkNode {
```

Comment:
You could even ditch the subclass and just register something like `static Result<ExecNode*> MakeTableConsumingSinkNode()` as a factory. There's no requirement that factories and nodes are 1:1, e.g. look at the scanner: https://github.com/apache/arrow/blob/09e4a5e04c9e39169b5264e0aa98a5cf0945ecba/cpp/src/arrow/dataset/scanner.cc#L838-L859

Reply:
I replaced it as a function.